07_errors/runnerdave/storer_test.go

```diff
@@ -0,0 +1,193 @@
+	testPuppy := puppy1()
+	cerr := s.store.CreatePuppy(&testPuppy)
+	r := require.New(s.T())
+	r.NoError(cerr, "Create should not produce an error")
+
+	// when
+	newPuppy, err := s.store.ReadPuppy(11)
+
+	// then
+	r.NoError(err, "Should be able to read a newly added puppy")
+	assert.Equal(testPuppy, newPuppy, "Newly added puppy should be equal to test puppy")
+}
+
+func (s *storerSuite) TestReadNonExistingPuppy() {
+	// given
+	assert := tassert.New(s.T())
+	r := require.New(s.T())
+
+	// when
+	_, err := s.store.ReadPuppy(12)
+
+	// then
+	r.Error(err, "Should produce an error when puppy is not found")
+	serr, ok := err.(*Error)
+	assert.True(ok)
+	assert.Equal(uint16(0x2), serr.Code)
+}
+
+func (s *storerSuite) TestDeletePuppy() {
+	// given
+	assert := tassert.New(s.T())
+	testPuppy := puppy1()
+	cerr := s.store.CreatePuppy(&testPuppy)
+	r := require.New(s.T())
+	r.NoError(cerr, "Create should not produce an error")
+
+	// when
+	err := s.store.DeletePuppy(11)
+
+	// then
+	r.NoError(err, "Should be able to delete a newly added puppy")
+	_, rerr := s.store.ReadPuppy(11)
+	serr, ok := rerr.(*Error)
+	assert.True(ok)
+	assert.Equal(uint16(0x2), serr.Code)
+}
+
+func (s *storerSuite) TestDeleteNonExistingPuppy() {
+	// given
+	assert := tassert.New(s.T())
+	r := require.New(s.T())
+
+	// when
+	err := s.store.DeletePuppy(11)
+
+	// then
+	r.Error(err, "Should not be able to delete a non existing puppy")
+	serr, ok := err.(*Error)
+	assert.True(ok)
+	assert.Equal(uint16(0x2), serr.Code)
+}
+
+func (s *storerSuite) TestCreateExistingPuppy() {
+	// given
+	assert := tassert.New(s.T())
+	testPuppy := puppy1()
+	cerr := s.store.CreatePuppy(&testPuppy)
+	r := require.New(s.T())
+	r.NoError(cerr, "Create should not produce an error")
+
+	// when
+	err := s.store.CreatePuppy(&testPuppy)
+
+	// then
+	assert.Error(err, "Should not be able to create twice a the same puppy")
+}
+
+func (s *storerSuite) TestCreatePuppyWithInvalidValue() {
+	// given
+	assert := tassert.New(s.T())
+	testPuppy := puppy3()
+	r := require.New(s.T())
+
+	// when
+	createError := s.store.CreatePuppy(&testPuppy)
+
+	// then
+	r.Error(createError, "Should not allow to create a puppy with invalid value")
+	serr, ok := createError.(*Error)
+	assert.True(ok)
+	assert.Equal(uint16(0x1), serr.Code)
+}
+
+func (s *storerSuite) TestUpdatePuppyWithInvalidValue() {
+	// given
+	assert := tassert.New(s.T())
+	testPuppy := puppy1()
+	updatePuppy := puppy3()
+	createError := s.store.CreatePuppy(&testPuppy)
+	r := require.New(s.T())
+	r.NoError(createError, "Create should not produce an error")
+
+	// when
+	uerr := s.store.UpdatePuppy(11, &updatePuppy)
+
+	// then
+	r.Error(uerr, "Should not allow to update a puppy with invalid value")
+	serr, ok := uerr.(*Error)
+	assert.True(ok)
+	assert.Equal(uint16(0x1), serr.Code)
+}
+
+func (s *storerSuite) SetupTest() {
+	s.store = s.storerType()
+}
+
+func TestStorer(t *testing.T) {
+	syncSuite := storerSuite{
+		store:      NewSyncStore(),
```

Comment:
Since `SetupTest()` sets `store` at the start of every test, this does not need to be initialised here - you can just leave it as the zero value (nil).